File: frontend/src/types/diary.ts
// 日记相关类型定义

export type EmotionTag = string
export type EventType = 'work' | 'relationship' | 'health' | 'achievement' | 'other'

export interface Diary {
  id: number
  user_id: number
  title: string
  content: string
  content_html?: string
  diary_date: string
  emotion_tags: EmotionTag[]
  importance_score: number
  word_count: number
  media_urls: string[]
  created_at: string
  updated_at: string
  is_analyzed: boolean
}

export interface DiaryCreate {
  title: string
  content: string
  content_html?: string
  diary_date?: string
  emotion_tags?: EmotionTag[]
  importance_score?: number
  media_urls?: string[]
}

export interface DiaryUpdate {
  title?: string
  content?: string
  content_html?: string
  diary_date?: string
  emotion_tags?: EmotionTag[]
  importance_score?: number
  media_urls?: string[]
}

export interface DiaryListResponse {
  items: Diary[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface TimelineEvent {
  id: number
  diary_id: number
  event_summary: string
  emotion_tag: string
  importance_score: number
  event_type: EventType
  related_entities: Record<string, any>
  event_date: string
  created_at: string
}

export interface EmotionStats {
  tag: string
  count: number
  percentage: number
}

export interface TerrainEvent {
  id: number | null
  diary_id: number
  summary: string
  emotion_tag: string | null
  importance_score: number
  event_type: EventType | null
  source_label?: string
}

export interface TerrainPoint {
  date: string
  energy: number | null
  valence: number | null
  density: number
  events: TerrainEvent[]
}

export interface TerrainPeak {
  date: string
  value: number
  label: string
  summary: string
}

export interface TerrainValley {
  date_range: [string, string]
  min_value: number
  days: number
  label: string
  summary: string
}

export interface TerrainInsights {
  peaks: TerrainPeak[]
  valleys: TerrainValley[]
  trend: 'ascending' | 'descending' | 'stable' | 'overall' | string
  trend_description: string
}

export interface TerrainMeta {
  start_date: string
  end_date: string
  total_events: number
  days_with_data: number
  total_days: number
}

export interface TerrainResponse {
  points: TerrainPoint[]
  insights: TerrainInsights
  meta: TerrainMeta
}

export interface DashboardInsightEmotion {
  tag: string
  label: string
  count: number
  percentage: number
}

export interface DashboardInsightDiary {
  id: number
  title: string
  diary_date: string
  emotion_tags: EmotionTag[]
  summary: string
  word_count: number
  importance_score: number
  is_analyzed: boolean
  analysis_path: string
}

export interface DashboardInsights {
  window_days: number
  generated_at: string
  stats: {
    total_diaries: number
    last_days_count: number
    this_month_count: number
    top_emotion: string
    top_emotion_label: string
    top_emotion_count: number
    trend: 'ascending' | 'descending' | 'stable' | string
    trend_label: string
    trend_delta: number
    risk_label: string
    risk_desc: string
    negative_ratio: number
  }
  ai_observation: {
    title: string
    summary: string
    encouragement: string
    source: string
  }
  emotion_stats: DashboardInsightEmotion[]
  insights: string[]
  recent_diaries: DashboardInsightDiary[]
  analysis_entry: {
    overall_path: string
    single_diary_path_template: string
    description: string
  }
}

export interface CareProgress {
  active_days: number
  actual_streak: number
  protected_streak: number
  shield_balance: number
  shielded_days: number
  shield_awarded: boolean
  shield_max: number
  weekly_goal: number
  weekly_active_count: number
  weekly_completed: boolean
  week_start: string
  today_status: 'ACTIVE' | 'REST' | 'SHIELDED' | 'MISSED' | 'PENDING' | string
  recent_statuses: Array<{
    date: string
    status: 'ACTIVE' | 'REST' | 'SHIELDED' | 'MISSED' | 'PENDING' | string
  }>
  message: string
}

export interface GrowthDailyInsight {
  date: string
  primary_emotion?: string
  summary?: string
  has_content: boolean
  cached: boolean
  source?: string
  message?: string
}

export interface PlanetEntry {
  emotion: string
  label: string
  planet: string
  emoji: string
  unlocked: boolean
  first_unlocked: string | null
  count: number
}

export interface PlanetCollection {
  total: number
  unlocked: number
  planets: PlanetEntry[]
}

export interface HeartLightCheckinPayload {
  emotion: string
  energy: number
  event?: string | null
  one_line_text?: string | null
  reflection_key?: string | null
  is_rest?: boolean
}

export interface HeartLightCheckin {
  id: number
  user_id: number
  checkin_date: string
  emotion: string
  energy: number
  event: string | null
  one_line_text: string | null
  reflection_key: string | null
  is_rest: boolean
  awarded_points: number
  created_at: string
  updated_at: string
}

export interface HeartLightCheckinSubmitResponse {
  checkin: HeartLightCheckin
  points_delta: number
  total_points: number
  new_planet: string | null
  streak: number
  shield_balance: number
}

export interface MonthCheckinDay {
  date: string
  emotion: string
  energy: number
  event: string | null
  is_rest: boolean
  one_line_excerpt: string | null
}

export interface MonthCheckinsResponse {
  month: string
  days: MonthCheckinDay[]
  diary_dates: string[]
  streak: number
  total_points: number
}

export type LightPointReason = 'checkin' | 'one_line' | 'planet_unlock' | string

export interface LightPointLedgerEntry {
  id: number
  delta: number
  reason: LightPointReason
  ref_date: string
  ref_id: number | null
  meta: Record<string, any> | null
  created_at: string
}

export interface LightPointsSummary {
  total: number
  recent_ledger: LightPointLedgerEntry[]
}

export interface LightPointByReason {
  checkin: number
  one_line: number
  planet_unlock: number
}

export interface TreasureTopDay {
  date: string | null
  points: number
}

export interface TreasureTopWeek {
  week_start: string | null
  points: number
}

export interface TreasureShield {
  balance: number
  max: number
  last_reward_streak: number
}

export interface TreasurePlanets {
  unlocked: number
  total: number
}

export interface TreasureResponse {
  total: number
  by_reason: LightPointByReason
  this_week_points: number
  this_month_points: number
  all_time_count: number
  top_day: TreasureTopDay
  top_week: TreasureTopWeek
  shield: TreasureShield
  planets: TreasurePlanets
  recent_ledger: LightPointLedgerEntry[]
}

File: backend/app/schemas/diary.py
"""
日记相关的 Pydantic Schemas
"""
from typing import Optional, List
from datetime import datetime, date
from pydantic import BaseModel, Field, field_validator


class DiaryCreate(BaseModel):
    """创建日记请求"""
    title: Optional[str] = Field(None, max_length=200, description="日记标题")
    content: str = Field(..., min_length=1, max_length=10000, description="日记内容")
    content_html: Optional[str] = Field(None, max_length=50000, description="日记HTML内容")
    diary_date: Optional[date] = Field(None, description="日记日期")
    emotion_tags: Optional[List[str]] = Field(None, description="情绪标签列表")
    importance_score: int = Field(default=5, ge=1, le=10, description="重要性评分（1-10）")
    images: Optional[List[str]] = Field(None, description="图片URL列表")

    @field_validator('diary_date', mode='before')
    @classmethod
    def set_default_date(cls, v):
        """如果没有提供日期，使用今天"""
        if v is None:
            return date.today()
        return v

    @field_validator('content')
    @classmethod
    def validate_content(cls, v):
        """验证内容不为空"""
        if not v or not v.strip():
            raise ValueError('日记内容不能为空')
        return v


class DiaryUpdate(BaseModel):
    """更新日记请求"""
    title: Optional[str] = Field(None, max_length=200, description="日记标题")
    content: Optional[str] = Field(None, min_length=1, max_length=10000, description="日记内容")
    content_html: Optional[str] = Field(None, max_length=50000, description="日记HTML内容")
    diary_date: Optional[date] = Field(None, description="日记日期")
    emotion_tags: Optional[List[str]] = Field(None, description="情绪标签列表")
    importance_score: Optional[int] = Field(None, ge=1, le=10, description="重要性评分（1-10）")
    images: Optional[List[str]] = Field(None, description="图片URL列表")
    is_analyzed: Optional[bool] = Field(None, description="是否已分析")


class DiaryResponse(BaseModel):
    """日记响应"""
    id: int
    user_id: int
    title: Optional[str]
    content: str
    content_html: Optional[str]
    diary_date: date
    emotion_tags: Optional[List[str]]
    importance_score: int
    word_count: int
    images: Optional[List[str]]
    is_analyzed: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class DiaryListResponse(BaseModel):
    """日记列表响应"""
    items: List[DiaryResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class TimelineEventCreate(BaseModel):
    """创建时间轴事件请求"""
    diary_id: Optional[int] = Field(None, description="关联的日记ID")
    event_date: date = Field(..., description="事件日期")
    event_summary: str = Field(..., min_length=1, max_length=500, description="事件摘要")
    emotion_tag: Optional[str] = Field(None, max_length=50, description="情绪标签")
    importance_score: int = Field(default=5, ge=1, le=10, description="重要性评分")
    event_type: Optional[str] = Field(None, max_length=50, description="事件类型")
    related_entities: Optional[dict] = Field(None, description="相关实体")


class TimelineEventResponse(BaseModel):
    """时间轴事件响应"""
    id: int
    user_id: int
    diary_id: Optional[int]
    event_date: date
    event_summary: str
    emotion_tag: Optional[str]
    importance_score: int
    event_type: Optional[str]
    related_entities: Optional[dict]
    created_at: datetime

    class Config:
        from_attributes = True


class HeartLightCheckinCreate(BaseModel):
    """心灯签到入参"""
    emotion: str = Field(..., max_length=20, description="情绪键：happy/calm/neutral/sad/anxious/angry/exhausted/rest")
    energy: int = Field(default=3, ge=1, le=5, description="能量 1-5")
    event: Optional[str] = Field(None, max_length=40, description="事件键：study/family/...")
    one_line_text: Optional[str] = Field(None, max_length=500, description="一句话心情")
    reflection_key: Optional[str] = Field(None, max_length=40, description="反思选项 key")
    is_rest: bool = Field(default=False, description="是否为休息日（今天不想写）")

    @field_validator("emotion")
    @classmethod
    def _strip_emotion(cls, v: str) -> str:
        v = (v or "").strip()
        if not v:
            raise ValueError("emotion 不能为空")
        return v

    @field_validator("one_line_text", mode="before")
    @classmethod
    def _normalize_one_line(cls, v):
        if v is None:
            return None
        v = str(v).strip()
        return v or None


class HeartLightCheckinOut(BaseModel):
    """心灯签到 DTO"""
    id: int
    user_id: int
    checkin_date: date
    emotion: str
    energy: int
    event: Optional[str]
    one_line_text: Optional[str]
    reflection_key: Optional[str]
    is_rest: bool
    awarded_points: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class LightPointLedgerOut(BaseModel):
    """映光流水条目"""
    id: int
    delta: int
    reason: str  # checkin / one_line / planet_unlock
    ref_date: date
    ref_id: Optional[int]
    meta: Optional[dict]
    created_at: datetime

    class Config:
        from_attributes = True


class LightPointsSummary(BaseModel):
    """映光总览"""
    total: int
    recent_ledger: List[LightPointLedgerOut]


class HeartLightCheckinSubmitResponse(BaseModel):
    """提交签到后的返回结构"""
    checkin: HeartLightCheckinOut
    points_delta: int
    total_points: int
    new_planet: Optional[str] = None
    streak: int
    shield_balance: int


class MonthCheckinDay(BaseModel):
    """月历单日"""
    date: date
    emotion: str
    energy: int
    event: Optional[str]
    is_rest: bool
    one_line_excerpt: Optional[str] = None


class MonthCheckinsResponse(BaseModel):
    """月历响应"""
    month: str  # YYYY-MM
    days: List[MonthCheckinDay]
    diary_dates: List[date]
    streak: int
    total_points: int


class LightPointByReason(BaseModel):
    """按来源汇总"""
    checkin: int = 0
    one_line: int = 0
    planet_unlock: int = 0


class TreasureTopDay(BaseModel):
    """映光峰值：某一天获得最多"""
    date: Optional[date] = None
    points: int = 0


class TreasureTopWeek(BaseModel):
    """映光峰值：某一周"""
    week_start: Optional[date] = None
    points: int = 0


class TreasureShield(BaseModel):
    balance: int
    max: int
    last_reward_streak: int


class TreasurePlanets(BaseModel):
    unlocked: int
    total: int


class TreasureResponse(BaseModel):
    """资产页聚合数据"""
    total: int
    by_reason: LightPointByReason
    this_week_points: int
    this_month_points: int
    all_time_count: int
    top_day: TreasureTopDay
    top_week: TreasureTopWeek
    shield: TreasureShield
    planets: TreasurePlanets
    recent_ledger: List[LightPointLedgerOut]
